### ISSUE TITLE
Get MechanicalObject from Python to C++

### ISSUE THREAD

**Issue by jjcasmar** [sofa-framework/SofaPython3#66]
I want to add some free function to retrieve the state from a MechanicalObject  in Python. I need to do it this way because my MechanicalObject is templated using a type that is not convertible to Python when wrapped on a sofa::Data (eigen matrix), so my idea is to add a free function which receives the MechanicalObject and extract the position from the data and returns a vector with it. 

Sadly, I am not sure which type of object my free function should receive. I have tried MechanicalObject<MyStateType> but that doesn't work. Also tried to receive a py::handle, but can't cast it to MechanicalObject<MyStateType>. Any idea on how to proceed?

**Comment by damienmarchal** [comment]
Hi @jjcasmar,

Suggestion to investigate
```cpp
py::object getPosition(BaseObject* object)   /// normally pybind will convert the py::object into a BaseObject
{
        auto customeType = dynamic_cast<MechanicalObject<Eigen::Matrix>*>(object); 
        if(!customType)
                /// Here rise a python type exception
        ///  here you can access the specific eigne matrix and return that something like that
        return toPythonEigen(customType.position);
}
```

**Comment by jjcasmar** [comment]
I actually did something very similar. Instead of received the BaseObject, Im receiving a py::handle which I cast to BaseObject and then dynamic_cast to MechanicalObject. Works fine so far. No need to toPythonEigen, I just created and `std::vector<Eigen::Matrix<double, 3, 3>>` and return that. Makes a copy but not an issue for now

I want to see if I can subclass MechanicalObject to provide class bindings directly, but so far, Im okay with the solution.
